lib/internal/modules/esm/fetch.js

```diff
@@ -0,0 +1,321 @@
+          }
+        }
+      }
+      if (res.statusCode > 303 || res.statusCode < 200) {
+        dispose();
+        reject(
+          new ERR_NETWORK_IMPORT_BAD_RESPONSE(
+            parsed.href,
+            'HTTP response returned status code of ' + res.statusCode));
+        return;
+      }
+      const { headers } = res;
+      const contentType = headers['content-type'];
+      if (!contentType) {
+        dispose();
+        reject(new ERR_NETWORK_IMPORT_BAD_RESPONSE(
+          parsed.href,
+          'the \'Content-Type\' header is required'));
+        return;
+      }
+      /**
+       * @type {CacheEntry}
+       */
+      const entry = {
+        resolvedHREF: parsed.href,
+        headers: {
+          'content-type': res.headers['content-type']
+        },
+        body: new Promise((f, r) => {
+          const buffers = [];
+          let size = 0;
+          let bodyStream = res;
+          let onError;
+          if (res.headers['content-encoding'] === 'br') {
+            bodyStream = createBrotliDecompress();
+            onError = function onError(error) {
+              bodyStream.close();
+              dispose();
+              reject(error);
+              r(error);
+            };
+            res.on('error', onError);
+            res.pipe(bodyStream);
+          } else if (res.headers['content-encoding'] === 'gzip' ||
+            res.headers['content-encoding'] === 'deflate') {
+            bodyStream = createUnzip();
+            onError = function onError(error) {
+              bodyStream.close();
+              dispose();
+              reject(error);
+              r(error);
+            };
+            res.on('error', onError);
+            res.pipe(bodyStream);
+          } else {
+            onError = function onError(error) {
+              dispose();
+              reject(error);
+              r(error);
+            };
+          }
+          bodyStream.on('error', onError);
+          bodyStream.on('data', (d) => {
+            ArrayPrototypePush(buffers, d);
+            size += d.length;
+          });
+          bodyStream.on('end', () => {
+            const body = entry.body = /** @type {Buffer} */(
+              BufferConcat(buffers, size)
+            );
+            f(body);
+          });
+        }),
+      };
+      cacheForGET.set(parsed.href, entry);
+      fulfill(entry);
+    });
+  });
+  cacheForGET.set(parsed.href, result);
+  return result;
+}
+
+/**
+ * Returns if an address has local status by if it is going to a local
+ * interface or is an address resolved by DNS to be a local interface
+ * @param {string} hostname url.hostname to test
+ * @returns {Promise<boolean>}
+ */
+async function isLocalAddress(hostname) {
+  try {
+    if (StringPrototypeStartsWith(hostname, '[') &&
+        StringPrototypeEndsWith(hostname, ']')) {
+      hostname = StringPrototypeSlice(hostname, 1, -1);
+    }
+    const addr = await dnsLookup(hostname, { verbatim: true });
```

Comment:
I think no because [net.isIP](https://nodejs.org/api/net.html#netisipinput) supports only IP addresses, not a hostname, and users will be using FQDNs not IP addresses:

```mjs
import net from 'net';
import dns from 'dns';

net.isIP('unpkg.com') // 0
dns.lookup('unpkg.com', { … }) // 2606:4700::6810:7caf 6
```

Reply:
Could the `net.BlockList` API be a better fit here?

Reply:
@jasnell I'd be more comfortable with an Accept list

Reply:
the `BlockList` can easily be used as an `Accept` list. Really, it's just an API for registering a number of IP addresses or ranges to test against. It can be used both for denying or granting.

```js
const list = new net.BlockList()
list.addAddress('::1', 'ipv6')
list.addRange('127.0.0.1', '127.255.255.255')
```

It has the added benefits of supporting a broad range of IPv4 and IPv6 formats... For instance, the added IPv4 range above would catch the IPv4-in-IPv6 syntax `::ffff:127.123.123.123` vs `:ffff:128.123.123.123`, whereas I believe the current logic could likely miss it.

Reply:
done